File: elixir/code_puppy_control/lib/code_puppy_control/config/tui.ex
defmodule CodePuppyControl.Config.TUI do
  @moduledoc """
  TUI theme, color, and display configuration.

  Manages banner colors, diff highlight colors, display suppression flags,
  and other UI-related settings from `puppy.cfg`.

  ## Config keys in `puppy.cfg`

  - `banner_color_<name>` — per-banner background color
  - `highlight_addition_color` — diff addition color
  - `highlight_deletion_color` — diff deletion color
  - `suppress_thinking_messages` — hide thinking/reasoning messages
  - `suppress_informational_messages` — hide info/success/warning messages
  - `grep_output_verbose` — show full grep output vs. concise
  - `diff_context_lines` — context lines for diff display (default 6)
  """

  alias CodePuppyControl.Config.Loader

  # ── Default banner colors ───────────────────────────────────────────────

  @default_banner_colors %{
    "thinking" => "deep_sky_blue4",
    "agent_response" => "medium_purple4",
    "shell_command" => "dark_orange3",
    "read_file" => "steel_blue",
    "edit_file" => "dark_goldenrod",
    "create_file" => "dark_goldenrod",
    "replace_in_file" => "dark_goldenrod",
    "delete_snippet" => "dark_goldenrod",
    "grep" => "grey37",
    "directory_listing" => "dodger_blue2",
    "agent_reasoning" => "dark_violet",
    "invoke_agent" => "deep_pink4",
    "subagent_response" => "sea_green3",
    "list_agents" => "dark_slate_gray3",
    "universal_constructor" => "dark_cyan",
    "terminal_tool" => "dark_goldenrod",
    "mcp_tool_call" => "dark_cyan",
    "shell_passthrough" => "medium_sea_green"
  }

  @doc """
  Get the color for a specific banner. Falls back to the default palette.
  """
  @spec banner_color(String.t()) :: String.t()
  def banner_color(name) do
    Loader.get_value("banner_color_#{name}") ||
      Map.get(@default_banner_colors, name, "blue")
  end

  @doc """
  Set the color for a specific banner.
  """
  @spec set_banner_color(String.t(), String.t()) :: :ok
  def set_banner_color(name, color) when is_binary(color) do
    CodePuppyControl.Config.Writer.set_value("banner_color_#{name}", color)
  end

  @doc """
  Return all banner colors (configured + defaults) as a map.
  """
  @spec all_banner_colors() :: %{String.t() => String.t()}
  def all_banner_colors do
    Map.new(@default_banner_colors, fn {name, default} ->
      {name, Loader.get_value("banner_color_#{name}") || default}
    end)
  end

  @doc """
  Reset a single banner color to its default.
  """
  @spec reset_banner_color(String.t()) :: :ok
  def reset_banner_color(name) do
    default = Map.get(@default_banner_colors, name, "blue")
    set_banner_color(name, default)
  end

  @doc """
  Reset all banner colors to defaults.
  """
  @spec reset_all_banner_colors() :: :ok
  def reset_all_banner_colors do
    Enum.each(@default_banner_colors, fn {name, color} ->
      set_banner_color(name, color)
    end)
  end

  @doc """
  Return the default banner colors map (read-only).
  """
  @spec default_banner_colors() :: %{String.t() => String.t()}
  def default_banner_colors, do: @default_banner_colors

  # ── Diff colors ─────────────────────────────────────────────────────────

  @doc "Return diff addition color (default `\"#0b1f0b\"`)."
  @spec diff_addition_color() :: String.t()
  def diff_addition_color do
    Loader.get_value("highlight_addition_color") || "#0b1f0b"
  end

  @doc "Set diff addition color."
  @spec set_diff_addition_color(String.t()) :: :ok
  def set_diff_addition_color(color) do
    CodePuppyControl.Config.Writer.set_value("highlight_addition_color", color)
  end

  @doc "Return diff deletion color (default `\"#390e1a\"`)."
  @spec diff_deletion_color() :: String.t()
  def diff_deletion_color do
    Loader.get_value("highlight_deletion_color") || "#390e1a"
  end

  @doc "Set diff deletion color."
  @spec set_diff_deletion_color(String.t()) :: :ok
  def set_diff_deletion_color(color) do
    CodePuppyControl.Config.Writer.set_value("highlight_deletion_color", color)
  end

  # ── Display flags ───────────────────────────────────────────────────────

  @doc "Return `true` if thinking messages are suppressed (default `false`)."
  @spec suppress_thinking?() :: boolean()
  def suppress_thinking?, do: truthy?("suppress_thinking_messages")

  @doc "Set suppress_thinking_messages."
  @spec set_suppress_thinking(boolean()) :: :ok
  def set_suppress_thinking(enabled) do
    CodePuppyControl.Config.Writer.set_value("suppress_thinking_messages", bool_str(enabled))
  end

  @doc "Return `true` if informational messages are suppressed (default `false`)."
  @spec suppress_informational?() :: boolean()
  def suppress_informational?, do: truthy?("suppress_informational_messages")

  @doc "Set suppress_informational_messages."
  @spec set_suppress_informational(boolean()) :: :ok
  def set_suppress_informational(enabled) do
    CodePuppyControl.Config.Writer.set_value("suppress_informational_messages", bool_str(enabled))
  end

  @doc "Return `true` if verbose grep output is enabled (default `false`)."
  @spec grep_output_verbose?() :: boolean()
  def grep_output_verbose?, do: truthy?("grep_output_verbose")

  @doc "Return diff context lines (default `6`, range `0–50`)."
  @spec diff_context_lines() :: non_neg_integer()
  def diff_context_lines do
    case Loader.get_value("diff_context_lines") do
      nil ->
        6

      val ->
        case Integer.parse(val) do
          {n, _} when n >= 0 and n <= 50 -> n
          _ -> 6
        end
    end
  end

  # ── Auto-save ───────────────────────────────────────────────────────────

  @doc "Return `true` if auto-save is enabled (default `true`)."
  @spec auto_save_session?() :: boolean()
  def auto_save_session?, do: truthy?("auto_save_session", true)

  @doc "Set auto_save_session."
  @spec set_auto_save_session(boolean()) :: :ok
  def set_auto_save_session(enabled) do
    CodePuppyControl.Config.Writer.set_value("auto_save_session", bool_str(enabled))
  end

  @doc "Return max saved sessions (default `20`)."
  @spec max_saved_sessions() :: non_neg_integer()
  def max_saved_sessions do
    case Loader.get_value("max_saved_sessions") do
      nil ->
        20

      val ->
        case Integer.parse(val) do
          {n, _} when n >= 0 -> n
          _ -> 20
        end
    end
  end

  # ── Private ─────────────────────────────────────────────────────────────

  @truthy_values MapSet.new(["1", "true", "yes", "on"])

  defp truthy?(key, default \\ false) do
    case Loader.get_value(key) do
      nil -> default
      val -> String.downcase(String.trim(val)) in @truthy_values
    end
  end

  defp bool_str(true), do: "true"
  defp bool_str(false), do: "false"
end

File: code_puppy/config/tui.py
"""TUI theme, color, and display configuration.

Mirrors ``CodePuppyControl.Config.TUI`` in the Elixir runtime.

Manages banner colors, diff highlight colors, display suppression flags,
and other UI-related settings from ``puppy.cfg``.

Config keys in puppy.cfg:

- ``banner_color_<name>`` — per-banner background color
- ``highlight_addition_color`` — diff addition color
- ``highlight_deletion_color`` — diff deletion color
- ``suppress_thinking_messages`` — hide thinking/reasoning messages
- ``suppress_informational_messages`` — hide info/success/warning messages
- ``grep_output_verbose`` — show full grep output vs. concise
- ``diff_context_lines`` — context lines for diff display (default 6)
"""

from __future__ import annotations

from code_puppy.config.loader import (
    _is_truthy,
    _make_bool_getter,
    _make_int_getter,
    _registered_cache,
    get_value,
    set_config_value,
)

__all__ = [
    "DEFAULT_BANNER_COLORS",
    "get_banner_color",
    "set_banner_color",
    "get_all_banner_colors",
    "reset_banner_color",
    "reset_all_banner_colors",
    "get_diff_addition_color",
    "set_diff_addition_color",
    "get_diff_deletion_color",
    "set_diff_deletion_color",
    "get_suppress_thinking_messages",
    "set_suppress_thinking_messages",
    "get_suppress_informational_messages",
    "set_suppress_informational_messages",
    "get_diff_context_lines",
    "get_grep_output_verbose",
]


# ---------------------------------------------------------------------------
# Default banner colors (jewel-tone palette with semantic meaning)
# ---------------------------------------------------------------------------

DEFAULT_BANNER_COLORS: dict[str, str] = {
    "thinking": "deep_sky_blue4",
    "agent_response": "medium_purple4",
    "shell_command": "dark_orange3",
    "read_file": "steel_blue",
    "edit_file": "dark_goldenrod",
    "create_file": "dark_goldenrod",
    "replace_in_file": "dark_goldenrod",
    "delete_snippet": "dark_goldenrod",
    "grep": "grey37",
    "directory_listing": "dodger_blue2",
    "agent_reasoning": "dark_violet",
    "invoke_agent": "deep_pink4",
    "subagent_response": "sea_green3",
    "list_agents": "dark_slate_gray3",
    "universal_constructor": "dark_cyan",
    "terminal_tool": "dark_goldenrod",
    "mcp_tool_call": "dark_cyan",
    "shell_passthrough": "medium_sea_green",
}


# ---------------------------------------------------------------------------
# Banner colors
# ---------------------------------------------------------------------------


def get_banner_color(banner_name: str) -> str:
    """Get the background color for a specific banner."""
    config_key = f"banner_color_{banner_name}"
    val = get_value(config_key)
    if val:
        return val
    return DEFAULT_BANNER_COLORS.get(banner_name, "blue")


def set_banner_color(banner_name: str, color: str) -> None:
    """Set the background color for a specific banner."""
    config_key = f"banner_color_{banner_name}"
    set_config_value(config_key, color)


def get_all_banner_colors() -> dict[str, str]:
    """Get all banner colors (configured or default)."""
    return {name: get_banner_color(name) for name in DEFAULT_BANNER_COLORS}


def reset_banner_color(banner_name: str) -> None:
    """Reset a banner color to its default."""
    default_color = DEFAULT_BANNER_COLORS.get(banner_name, "blue")
    set_banner_color(banner_name, default_color)


def reset_all_banner_colors() -> None:
    """Reset all banner colors to their defaults."""
    for name, color in DEFAULT_BANNER_COLORS.items():
        set_banner_color(name, color)


# ---------------------------------------------------------------------------
# Diff colors
# ---------------------------------------------------------------------------


@_registered_cache
def get_diff_addition_color() -> str:
    """Get the base color for diff additions. Default: ``#0b1f0b``."""
    val = get_value("highlight_addition_color")
    if val:
        return val
    return "#0b1f0b"


def set_diff_addition_color(color: str) -> None:
    """Set the color for diff additions."""
    set_config_value("highlight_addition_color", color)


@_registered_cache
def get_diff_deletion_color() -> str:
    """Get the base color for diff deletions. Default: ``#390e1a``."""
    val = get_value("highlight_deletion_color")
    if val:
        return val
    return "#390e1a"


def set_diff_deletion_color(color: str) -> None:
    """Set the color for diff deletions."""
    set_config_value("highlight_deletion_color", color)


# ---------------------------------------------------------------------------
# Diff context lines
# ---------------------------------------------------------------------------


get_diff_context_lines = _make_int_getter(
    "diff_context_lines",
    default=6,
    min_val=0,
    max_val=50,
    doc="""Return the number of context lines for diff display (default 6, range 0-50).""",
)


# ---------------------------------------------------------------------------
# Display suppression flags
# ---------------------------------------------------------------------------


get_suppress_thinking_messages = _make_bool_getter(
    "suppress_thinking_messages",
    default=False,
    doc="""Return True if thinking messages are suppressed (default False).""",
)


def set_suppress_thinking_messages(enabled: bool) -> None:
    """Set suppress_thinking_messages."""
    set_config_value("suppress_thinking_messages", "true" if enabled else "false")


def get_suppress_informational_messages() -> bool:
    """Return True if informational messages are suppressed (default False)."""
    return _is_truthy(get_value("suppress_informational_messages"), default=False)


def set_suppress_informational_messages(enabled: bool) -> None:
    """Set suppress_informational_messages."""
    set_config_value("suppress_informational_messages", "true" if enabled else "false")


# ---------------------------------------------------------------------------
# Grep output verbosity
# ---------------------------------------------------------------------------


get_grep_output_verbose = _make_bool_getter(
    "grep_output_verbose",
    default=False,
    doc="""Return True if verbose grep output is enabled (default False).""",
)
